### PR TITLE
[Feature][Connector-V2][File] Add option and factory for file connectors

### LINKED ISSUE #3367: [Feature][Connector V2] expose configurable options in Hive

### LINKED ISSUE #3368: [Feature][Connector V2] expose configurable options in FileLocal @TyrantLucifer

### LINKED ISSUE #3369: [Feature][Connector V2] expose configurable options in FileHadoop @TyrantLucifer

### LINKED ISSUE #3370: [Feature][Connector V2] expose configurable options in FileFtp @TyrantLucifer

### LINKED ISSUE #3371: [Feature][Connector V2] expose configurable options in FileSftp @TyrantLucifer

### LINKED ISSUE #3372: [Feature][Connector V2] expose configurable options in S3 @TyrantLucifer

### LINKED ISSUE #3373: [Feature][Connector V2] expose configurable options in OSS @TyrantLucifer

### PR DESCRIPTION
**Pull request by TyrantLucifer**
## Purpose of this pull request

close #3367 
close #3368 
close #3369 
close #3370 
close #3371 
close #3372 
close #3373 

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

**Comment by TyrantLucifer** [comment]
@Hisoka-X @EricJoy2048 PTAL.

File: docs/en/connector-v2/sink/Hive.md [hunk 1]
```diff
@@ -24,15 +24,11 @@ By default, we use 2PC commit to ensure `exactly-once`
 
 ## Options
 
-| name                  | type   | required                                    | default value                                                 |
-|-----------------------| ------ |---------------------------------------------| ------------------------------------------------------------- |
-| table_name            | string | yes                                         | -                                                              |
-| metastore_uri         | string | yes                                         | -                                                              |
-| partition_by          | array  | required if hive sink table have partitions | -                                                             |
```
Comment (EricJoy2048) [inline docs/en/connector-v2/sink/Hive.md:31]: Why remove this parameter?

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import org.apache.seatunnel.common.utils.TimeUtils;
+
+import java.util.List;
+
+public class BaseSinkConfig {
+    public static final String SEATUNNEL = "seatunnel";
+    public static final String NON_PARTITION = "NON_PARTITION";
+    public static final String TRANSACTION_ID_SPLIT = "_";
+    public static final String TRANSACTION_EXPRESSION = "transactionId";
+    public static final Option<String> COMPRESS_CODEC = Options.key("compress_codec")
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java:39]: Please add this parameter to sink doc.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            .stringType()
+            .noDefaultValue()
+            .withDescription("Compression codec");
+    public static final Option<String> DATE_FORMAT = Options.key("date_format")
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]: same as above.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            .stringType()
+            .defaultValue(DateUtils.Formatter.YYYY_MM_DD.getValue())
+            .withDescription("Date format");
+    public static final Option<String> DATETIME_FORMAT = Options.key("datetime_format")
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]: same as above.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            .stringType()
+            .defaultValue(DateTimeUtils.Formatter.YYYY_MM_DD_HH_MM_SS.getValue())
+            .withDescription("Datetime format");
+    public static final Option<String> TIME_FORMAT = Options.key("time_format")
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]: same as above.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            .stringType()
+            .defaultValue(TimeUtils.Formatter.HH_MM_SS.getValue())
+            .withDescription("Time format");
+    public static final Option<String> FILE_PATH = Options.key("path")
+            .stringType()
+            .noDefaultValue()
+            .withDescription("File path");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]:
```suggestion
            .withDescription("The target dir path");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    public static final Option<String> FIELD_DELIMITER = Options.key("field_delimiter")
+            .stringType()
+            .defaultValue(String.valueOf(String.valueOf('\001')))
+            .withDescription("Field delimiter");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]:
```suggestion
            .withDescription("The separator between columns in a row of data. Only needed by `text` and `csv` file format.");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    public static final Option<String> ROW_DELIMITER = Options.key("row_delimiter")
+            .stringType()
+            .defaultValue("\n")
+            .withDescription("Row delimiter");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]:
```suggestion
            .withDescription("The separator between rows in a file. Only needed by `text` and `csv` file format.");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    public static final Option<List<String>> PARTITION_BY = Options.key("partition_by")
+            .listType()
+            .noDefaultValue()
+            .withDescription("Partition keys list");
+    public static final Option<String> PARTITION_DIR_EXPRESSION = Options.key("partition_dir_expression")
+            .stringType()
+            .defaultValue("${k0}=${v0}/${k1}=${v1}/.../${kn}=${vn}/")
+            .withDescription("Partition directory expressions");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]:
```suggestion
            .withDescription("If the `partition_by` is specified, we will generate the corresponding partition directory based on the partition information, and the final file will be placed in the partition directory. Default `partition_dir_expression` is `${k0}=${v0}/${k1}=${v1}/.../${kn}=${vn}/`. `k0` is the first partition field and `v0` is the value of the first partition field.");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            .defaultValue(false)
+            .withDescription("Whether to write partition fields to file");
+    public static final Option<String> TMP_PATH = Options.key("tmp_path")
+            .stringType()
+            .defaultValue("/tmp/seatunnel")
+            .withDescription("Data write temporary path");
+    public static final Option<String> FILE_NAME_EXPRESSION = Options.key("file_name_expression")
+            .stringType()
+            .defaultValue("${transactionId}")
+            .withDescription("Describe the file expression which will be created into the path");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]:
```suggestion
            .withDescription("`file_name_expression` describes the file expression which will be created into the `path`. We can add the variable `${now}` or `${uuid}` in the `file_name_expression`, like `test_${uuid}_${now}`,`${now}` represents the current time, and its format can be defined by specifying the option `filename_time_format`.");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSourceConfig.java [hunk 11]
```diff
@@ -17,13 +17,39 @@
...
+            .defaultValue(DateTimeUtils.Formatter.YYYY_MM_DD_HH_MM_SS.getValue())
+            .withDescription("Datetime format");
+    public static final Option<String> TIME_FORMAT = Options.key("time_format")
+            .stringType()
+            .defaultValue(TimeUtils.Formatter.HH_MM_SS.getValue())
+            .withDescription("Time format");
+    public static final Option<Boolean> PARSE_PARTITION_FROM_PATH = Options.key("parse_partition_from_path")
+            .booleanType()
+            .defaultValue(true)
+            .withDescription("Is or not parse partition fields from file path");
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSourceConfig.java]:
```suggestion
            .withDescription("Wether parse partition fields from file path");
```

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSourceConfig.java [hunk 12]
```diff
@@ -17,13 +17,39 @@
...
 public class BaseSourceConfig {
-    public static final String FILE_TYPE = "type";
-    public static final String FILE_PATH = "path";
-    public static final String SCHEMA = "schema";
-    public static final String DELIMITER = "delimiter";
-    public static final String DATE_FORMAT = "date_format";
-    public static final String DATETIME_FORMAT = "datetime_format";
-    public static final String TIME_FORMAT = "time_format";
-    public static final String PARSE_PARTITION_FROM_PATH = "parse_partition_from_path";
+    public static final Option<String> FILE_TYPE = Options.key("type")
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSourceConfig.java:27]: Use Enum better. We need let web ui know the value list.

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            .stringType()
+            .noDefaultValue()
+            .withDescription("The file path of target files");
+    public static final Option<String> FIELD_DELIMITER = Options.key("field_delimiter")
+            .stringType()
+            .defaultValue(String.valueOf(String.valueOf('\001')))
+            .withDescription("The separator between columns in a row of data. Only needed by `text` and `csv` file format");
+    public static final Option<String> ROW_DELIMITER = Options.key("row_delimiter")
+            .stringType()
+            .defaultValue("\n")
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]: How about defining a constant？

File: seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+                    "and the final file will be placed in the partition directory. " +
+                    "Default `partition_dir_expression` is `${k0}=${v0}/${k1}=${v1}/.../${kn}=${vn}/`. " +
+                    "`k0` is the first partition field and `v0` is the value of the first partition field.");
+    public static final Option<Boolean> IS_PARTITION_FIELD_WRITE_IN_FILE = Options.key("is_partition_field_write_in_file")
+            .booleanType()
+            .defaultValue(false)
+            .withDescription("Whether to write partition fields to file");
+    public static final Option<String> TMP_PATH = Options.key("tmp_path")
+            .stringType()
+            .defaultValue("/tmp/seatunnel")
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/config/BaseSinkConfig.java]: Same as above.

**Review by ashulin — APPROVED** [review]
LGTM

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**